core/samples/data/ser/json/main.cpp

```diff
@@ -0,0 +1,159 @@
+#include <cubos/core/log.hpp>
+#include <cubos/core/memory/stream.hpp>
+#include <cubos/core/data/ser/serializer.hpp>
+
+using cubos::core::memory::Stream;
+using cubos::core::data::Serializer;
+using cubos::core::reflection::Type;
+
+/// [JSON package]
+#include <nlohmann/json.hpp>
+
+/// [Create objects that will hold the data]    
+auto jsonObject = nlohmann::json::object();
+auto jsonArray = nlohmann::json::array();
+auto jsonValue = nlohmann::json::object();
+
+/// [MACRO to add hooks for primitive types]
+#define AUTO_HOOK(type, ...)                                                \
+    this->hook<type>([](Serializer&, const Type&, const void* value) {      \
+        jsonValue = *static_cast<const type*>(value);                       \
+        return true;                                                        \
+    })
+
+/// [JSON serializer]
+class JSONSerializer : public Serializer
+{
+public:
+    JSONSerializer();
+    std::string output() const {
+        return json.dump();
+    }
+
+protected:
+    bool decompose(const Type& type, const void* value) override;
+
+private:
+    nlohmann::json json;
+};
+/// [JSON serializer]
```

Comment:
Instead of declaring this here, this should be in a new file `include/cubos/core/data/ser/json.hpp`. You can copy it from DebugSerializer's `debug.hpp`.